src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -1038,37 +1030,45 @@ func (i *fetchTaggedResultsIter) Next(ctx context.Context) bool {
 
 // acquire a block permit for a series ID. returns true if a permit is available.
 func (i *fetchTaggedResultsIter) acquire(ctx context.Context, idx int) (bool, error) {
-	if i.blocksAvailable > 0 {
-		i.blocksAvailable--
-	} else {
+	var curPermit permits.Permit
+	if len(i.permits) > 0 {
+		curPermit = i.permits[len(i.permits)-1]
+	}
+	if curPermit == nil || curPermit.QuotaRemaining() <= 0 {
 		if i.idx == idx {
 			// block acquiring if we need the block readers to fulfill the current fetch.
-			if err := i.blockPermits.Acquire(ctx); err != nil {
+			permit, err := i.blockPermits.Acquire(ctx)
+			if err != nil {
```

Comment:
interesting; so if we do run into an error, we don't technically release all held permits? Or is that handled elsewhere? (separate comment from your change I think)

Reply:
when an error happens, the fetched tagged iterator is closed, which releases all held permits.